(DOCSP-21149): Refactor to improve readability from grade C to B, or grade 12 to 8 (#1863)
mongodb/docs-realm: source/sdk/java/examples/reset-a-client-realm.txt

   To learn about the causes of and strategies for handling client
   resets, check out the Sync :ref:`Client Resets <client-resets>` page.

> The realm file in your client handles reads and writes on the device. When 

Reviewer: Technically the SDK handles reads and writes (is it ever correct to say that a file "handles reads and writes"? I'm not sure this sentence means anything in its current state. Add in the passive wording beneath and I suspect this paragraph was better in its original form. Maybe we could just break up the original second sentence into two sentences instead?

Author: Your point is fair. But I think as-written, we're missing an opportunity to better set up this problem to make it easier or more intuitive for devs to understand the concept of a local realm having changes that don't exist in the Cloud. I'll have a think about this intro.

Author: What if I change the first line in this paragraph to:
> The SDK reads and writes to a realm file on the device.

Then the opening paragraph becomes:
> The SDK reads and writes to a realm file on the device. When 
you use Realm Sync, this local realm syncs with the application backend. 
Some conditions can cause the realm to be unable to sync with the backend. 
When this occurs, you get a **client reset error**.

you use Realm Sync, this local realm syncs with the application backend. 
Some conditions can cause the realm to be unable to sync with the backend. 
When this occurs, you get a **client reset error**. 

This error means you must reset the realm file in the client application. 
Clients in this state may continue to run and save data locally. Until you 

> perform the client reset, the realm does not sync with the backend.

Reviewer: I like this rewording, definitely an improvement from the original wording. Probably a good idea to split this into two paragraphs anyway, where one talks about what a client reset is and the other discusses the impact a client reset has on your application.

Choose a **client reset strategy** to handle client reset errors. These 
strategies restore realm to a syncable state, but have tradeoffs:

- :ref:`Discard Unsynced Changes <java-discard-unsynced-changes>`. Restore 

>   Sync by discarding local changes since the last sync. Maintains change 

Reviewer: This wording is great, huge improvement!

  listeners.
- :ref:`Manually Recover Unsynced Changes 
  <java-manually-recover-unsynced-changes>`:. Move the unsyncable realm and 
  download a new copy. Invalidates change listeners.

Both options let you write custom logic to recover local changes. Neither 
option can recover local changes for you. 

> Discard unsynced changes is an improved version of manual recovery. However, 

Reviewer: "Improved" is a little subjective here -- maybe, "easier to implement" would be more accurate? (and a reword to support that phrasing)

Author: What about:
> Discard unsynced changes is a less complex alternative to manual recovery.

Reviewer: LGTM

.. versionadded:: 10.10.0

The **discard unsynced changes** performs client resets "seamlessly." This 
strategy requires minimal code and minimal disruption. This strategy 

> does not close the realm or miss notifications.

Reviewer: I think I prefer the original first sentence to the new first two sentences, the new third sentence is a big improvement though. Personally I feel that the new first sentence has no real meaning to the reader and we should just trim it.

Author: The original first sentence is _long_ and difficult to parse. I'm not tied to this rewording specifically, but I do think we need to break that sentence up and make it easier to parse. If you want to wordsmith it a bit with Hemingway or Readable, I'm happy to go with a different wording that you prefer. Or I can take another shot at this and try to find an alternative wording you'll like better.

(That "seamless" thing was a nod to our product positioning for this feature, but yeah, it doesn't actually communicate effectively to the developer.)

Reviewer: Added a suggested wording, I'm not 100% happy with it but maybe you can take some inspiration.

> The **discard unsynced changes** performs client resets "seamlessly." This 
> strategy requires minimal code and minimal disruption. This strategy 
> does not close the realm or miss notifications.

Reviewer: How about:

```suggestion
**Discard unsynced changes** is a client reset strategy provided by the SDK.
This strategy requires minimal code. This strategy does not close the
realm or miss notifications.
```

It *does* delete all local changes made since the last successful sync. 

> Do not use this strategy if your application cannot lose unsynced data. 

Reviewer: The SDK team specifically asked for the "data already written to the client realm file but not yet synced to the backend" clarification, IIRC -- "unsynced data" is not necessarily a clear concept for readers. I would probably add a sentence that clarifies that "unsynced data == local changes made since the last successful sync".

Author: That's fair - the wording was just awkward and difficult to parse. This is part of what I was trying to solve with the intro rewording - trying to establish that the SDK persists changes to the local realm, and then Sync attempts to apply these changes to the data in the cloud, which I think makes the concept of unsynced data clearer. I'll have another think about how we might communicate this more clearly.

This strategy provides a frozen copy of the realm from before reset. You 
can attempt to recover changes from this frozen realm. However, the code 
required for a manual recovery is quite complex. This is not recommended 

> for most applications.

Reviewer: I believe I omitted this detail originally because the "frozen copy" is impermanent and not reliable for a real recovery operation, so users can't really rely on it for anything. As a result it's easier and simpler to leave it out of the explanation since it just confuses things (IMO)

Author: When I was working on the Swift SDK version of this doc, the team thought it was important to call this out as it gives devs an _option_ to do something with the data if they want it, whereas they can't do that if they don't know it exists. Can this frozen realm not be copied somewhere more permanent and used for a recovery operation? That was the suggestion in the Swift SDK.

Reviewer: Interesting, as far as I know you can't copy the "frozen" instance of the realm because it only exists as a diff state in the realm file itself. I say it's impermanent because if your application crashes, or gets closed by the user, or the device battery dies, or whatever... you'll never be able to access that frozen instance again. So there's no guarantee you can do anything with it. Probably not a big deal to mention it here, I didn't think it had much value for developers though.

Author: Hmm. Removing this paragraph, then. If your SDK team didn't feel it was important to mention, and you don't think it provides much value to developers, then I suppose there's no need to mention it - better to not give developers the wrong impression about what they can do with this client reset stuff.

Discard unsynced changes cannot handle :ref:`breaking or destructive schema 

> changes <destructive-changes-synced-schema>`. With breaking changes, this 

Reviewer: ```suggestion
changes <destructive-changes-synced-schema>`. When breaking changes occur, this 
```
I don't think there's any actual grammatical issue here, but for some reason "With breaking changes" just sounds... strange? to me. Take it or leave it.

Author: That's fair. I did try that initially but it made the sentence "too long." What about:

> When breaking changes occur, the SDK falls back to manual recovery mode.

strategy falls back to manual recovery mode.

To use this strategy, pass an instance of 
:java-sdk:`DiscardUnsyncedChangesStrategy
<io/realm/mongodb/sync/DiscardUnsyncedChangesStrategy.html>` to the
:java-sdk:`defaultSyncClientResetStrategy()
<io/realm/mongodb/AppConfiguration.Builder.html#defaultSyncClientResetStrategy(io.realm.mongodb.sync.DiscardUnsyncedChangesStrategy)>`
builder method when you instantiate your ``App``. Your
``DiscardUnsyncedChangesStrategy`` instance must implement the following
methods:

> - ``onBeforeReset()``. Realm calls this block when it receives a client 

Reviewer: ```suggestion
- ``onBeforeReset()``. The SDK calls this block when it receives a client 
```
I try to avoid using "Realm" as a subject for active verbs since realm (the file), Realm (the backend), Realm (the company), etc. are really easy to confuse. I'd suggest "Realm Database" here instead, but that gets kind of repetitive in two bullet points in a row.

Reviewer: Additionally, I think it's useful to clarify that this method call happens *BEFORE* any recovery logic runs. Maybe we can add that in somewhere.

  reset error from the backend.
- ``onAfterReset()``. Realm calls this block after successfully executing 
  this strategy. This block provides a frozen copy of the original realm. 

>   It also returns a live instance of the realm in a syncable state.

Reviewer: I like this, aside from the Realm active subjectification I mentioned in the previous comment.

> - ``onError()``: Mimics the "manually recover unsynced changes" strategy. 

Reviewer: Perhaps instead of "mimics" we can point to the specific method in "manually recover unsynced changes" that this behaves just like.

Author: Is that this: [defaultClientResetStrategy](https://www.mongodb.com/docs/realm-sdks/java/latest/io/realm/mongodb/AppConfiguration.Builder.html?_ga=2.12531955.691438089.1650290917-78975312.1612214951#defaultSyncClientResetStrategy(io.realm.mongodb.sync.ManuallyRecoverUnsyncedChangesStrategy))?

Reviewer: Yep!

.. include:: /includes/destructive-schema-change-app-update.rst

The discard unsynced changes strategy cannot handle breaking changes. You 
must manually handle the client reset in the ``onError()`` method. This 

> example discards unsynced changes to handle the client reset:

Reviewer: ```suggestion
example manually discards unsynced changes to handle the client reset:
```

I'd specifically call out that this is a *manual* process, as opposed to the automatically discard strategy this happens inside of. Otherwise nice work streamlining this!
